### PR TITLE
Implement retrofit and coroutine to retrieve data from unsplash api

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -23,15 +32,66 @@ android {
...
+
+    // Core with ktx
+    implementation 'androidx.core:core-ktx:1.3.0'
+
+    // Support
+    implementation 'androidx.appcompat:appcompat:1.1.0'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
+
+    // Material Design
+    implementation 'com.google.android.material:material:1.2.0-beta01'
```
Comment (hua9495) [inline app/build.gradle]: Avoid using beta version, unless there's a feature you really need in the beta version

File: build.gradle [hunk 2]
```diff
@@ -20,7 +25,7 @@ allprojects {
     repositories {
         google()
         jcenter()
-        
+        maven { url  "https://dl.bintray.com/unsplash/unsplash-photopicker-android" }
```
Comment (hua9495) [inline build.gradle]: Guess this should be unused, remove it

File: app/src/main/java/com/alexchan/wallpaper/ui/wallpaper/main_new/MainNewViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    
+    val status: LiveData<UnsplashApiStatus>
+        get() = _status
+
+    val photoProperties: LiveData<List<Photo>>
+        get() = _photoProperties
+
+    private var viewModelJob = Job()
+
+    private val coroutineScope = CoroutineScope(viewModelJob + Dispatchers.Main)
```
Comment (hua9495) [inline app/src/main/java/com/alexchan/wallpaper/ui/wallpaper/main_new/MainNewViewModel.kt]: use `viewModelScope` instead

**Review by hua9495 — COMMENTED** [review]
Need some clarification on what is Data Binding Library and why you use it. Also, no need to create new PR for each push, delete the old one if you really need to.

File: app/src/main/java/com/alexchan/wallpaper/ui/wallpaper/dashboard/DashboardFragment.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.ViewModelProvider
+import com.alexchan.wallpaper.adapter.wallpaper.PhotoGridAdapter
+import com.alexchan.wallpaper.databinding.FragmentDashboardBinding
+
+class DashboardFragment : Fragment() {
+
+    private val viewModel: MainNewViewModel by lazy {
+        ViewModelProvider(this).get(MainNewViewModel::class.java)
```
Comment (hua9495) [inline app/src/main/java/com/alexchan/wallpaper/ui/wallpaper/dashboard/DashboardFragment.kt]: Rename this viewmodel

**Review by hua9495 — APPROVED** [review]

**Comment by ghost** [comment]
### Dashboard


### Wallpaper Details 1


### Wallpaper Details 2 - with over a thousand liked


### Wallpaper Details 3 - with null location

**Comment by ghost** [comment]
### Dashboard - new tab indicator


### Wallpaper Details - new image indicator

**Comment by ghost** [comment]
### Dashboard - new theme


### Wallpaper Details - new theme, title, over a thousand photos

**Comment by ghost** [comment]
### Dashboard - staggered grid layout manager


### Wallpaper Details - staggered grid layout manager

**Comment by ghost** [comment]
### Dashboard - search function


### Dashboard - change display view type function

File: app/src/main/java/com/alexchan/wallpaper/ui/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        Log.d(TAG, supportFragmentManager.findFragmentById(R.id.mainNavHostFragment)
+            ?.findNavController()?.currentDestination?.id.toString())
+        when (supportFragmentManager.findFragmentById(R.id.mainNavHostFragment)?.findNavController()?.currentDestination?.id) {
+            R.id.wallpaperFragment -> navigationView.setCheckedItem(R.id.wallpaperFragment)
+            R.id.notificationFragment -> navigationView.setCheckedItem(R.id.notificationFragment)
+            R.id.profileFragment -> navigationView.setCheckedItem(R.id.profileFragment)
+        }
+    }
+
+    override fun onBackPressed() {
```
Comment (hua9495) [inline app/src/main/java/com/alexchan/wallpaper/ui/MainActivity.kt:69]: This is actually bad. If you have a toolbar in the details screen, try this `toolbar.setupWithNavController(findNavController())`

File: app/src/main/java/com/alexchan/wallpaper/ui/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        val searchView = topToolbar.menu.findItem(R.id.search).actionView as SearchView
+        val navHostFragment = supportFragmentManager.findFragmentById(R.id.mainNavHostFragment)
+        val navController = navHostFragment?.findNavController()
+        val navCurrentDestination = navController?.currentDestination?.id
+        val count = supportFragmentManager.backStackEntryCount
+        Log.d(TAG, count.toString())
+        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
```
Comment (hua9495) [inline app/src/main/java/com/alexchan/wallpaper/ui/MainActivity.kt]: if you really want to do this, replace `if` with `when`.

File: app/src/main/java/com/alexchan/wallpaper/ui/wallpaper/dashboard/details/WallpaperDetailsViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        Log.d("Display Liked", it.user!!.totalLikes.toString())
+        app.applicationContext.getString(
+            when (it.user.isTotalLikesOverThousand) {
+                true -> R.string.userTotalThousandOverLiked
+                false -> R.string.userTotalLiked
+        }, when (it.user.isTotalLikesOverThousand) {
+                true -> {
+                    val df = DecimalFormat("#.#")
+                    df.roundingMode = RoundingMode.FLOOR
+                    Log.d("Display Calculation", df.format(((it.user!!.totalLikes).toFloat() / 1000)))
```
Comment (hua9495) [inline app/src/main/java/com/alexchan/wallpaper/ui/wallpaper/dashboard/details/WallpaperDetailsViewModel.kt]: Code is repeating, should place it in one place.

File: app/src/main/res/layout/fragment_wallpaper.xml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    android:layout_marginTop="?attr/actionBarSize"
+    android:orientation="vertical">
+
+    <com.google.android.material.tabs.TabLayout
+        android:id="@+id/tabLayout"
+        android:layout_width="match_parent"
+        android:layout_height="?attr/actionBarSize"
+        android:theme="@style/TabSelectedTheme"
+        app:tabIndicator="@drawable/selected_dot"
+        app:tabIndicatorColor="#f44336"
```
Comment (hua9495) [inline app/src/main/res/layout/fragment_wallpaper.xml]: Best practices, avoid using hex color directly, group all in `colors.xml`

File: app/src/main/res/layout/fragment_wallpaper_details.xml [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            app:elevation="0dp">
+
+            <com.google.android.material.appbar.MaterialToolbar
+                android:id="@+id/titleTopToolbar"
+                android:layout_width="match_parent"
+                android:layout_height="?android:attr/actionBarSize"
+                app:title="@{wallpaperDetailsViewModel.userPhotoCollection.user.name}"
+                app:titleTextColor="#8d8d8d"
```
Comment (hua9495) [inline app/src/main/res/layout/fragment_wallpaper_details.xml]: same as above

**Comment by hua9495** [comment]
This PR is getting bigger, amend with the comments above and merge this PR. Continue with another PR.

**Comment by ghost** [comment]
I'm unable to merge this PR, it displays the following message.

```
This branch has no conflicts with the base branch
Only those with write access to this repository can merge pull requests.
```

**Comment by hua9495** [comment]
Go to email accept the invitation, and try again

**Pull request merged**